### ISSUE TITLE
GNU readline and onefile

### ISSUE THREAD

**Issue by jayvdb** [Nuitka/Nuitka#1092]
Nuitka includes GNU readline by default, and this has legal implications for use with onefile mode.

https://github.com/pyinstaller/pyinstaller/issues/4471 goes into the details well enough, and I suggest a similar approach for Nuitka: a very strong warning in the Nuitka stdout/stderr, or even better an explicit flag that the user must enable before readline will be included, at least in onefile mode.

Even better would be to use Pythons from https://python-build-standalone.readthedocs.io/en/latest/ , and include their libedit Python module instead of the readline Python module.  This isnt a perfect solution, as many Python packages only expect libedit on macOS, but well maintained TUI libraries will be able to utilise libedit on Windows and Linux and other platforms, so the user experience wont be effected significantly.

**Comment by kayhayen** [comment]
There is going to be a fork of CPython for Nuitka, called Nuitka-Python, which should address the issue. That is going to lend itself to static linking and integrate better with Nuitka overall.

I am not a lawer. 

Obviously the code in question is inactive in all deployments. Also, I am confused as to how standalone it would be OK, and how putting into onefile would not be. Putting something in an self extracting archive, doesn't really change things, does it? Mind you, no source code was used of the DLL, so how could the linking be considered different? However, I will take it as an indication, that this should be removed from the distribution, unless of course this prevents the Python DLL from loading. Then we would have to stub it or something.

**Comment by jayvdb** [comment]
A self-extracting archive may trigger the same problem, but not extraction is once-off and the execution occurs from the directory after the extraction is done.  The GPL aspect that causes this problem is whether the user has the ability to replace the libreadline with their own - if they cant easily do this, it is no longer an external runtime dependency.

**Comment by kayhayen** [comment]
@jayvdb i see, totally makes sense

**Comment by tropxy** [comment]
> A self-extracting archive may trigger the same problem, but not extraction is once-off and the execution occurs from the directory after the extraction is done. The GPL aspect that causes this problem is whether the user has the ability to replace the libreadline with their own - if they cant easily do this, it is no longer an external runtime dependency.

Hi, I think that is specially true for a LGPL license, but in this case it is a GPL one and as far as I know, even just having a shared linking to the library triggers the Disclose source clause: https://choosealicense.com/licenses/gpl-3.0/

Also check:
https://www.legal.io/articles/5170736/Open-Source-Software-a-legal-guide

**Comment by kayhayen** [comment]
The Nuitka onefile code is available in all its contents, so not being a lawyer, I feel no danger by using it to distribute a file. Are you saying, that say using `rar` to pack/unpack a GPLed binary has any effect on the other binaries contained? The LGPL would indeed come into play if it was statically linked.

That said, the onefile packing/unpacking is GPL compatible in nature due to the Apache license, it would even be allowed to link itself to readline, which it does not.

Unlike pyinstaller I am willing to discuss this, but honestly I fail to see the point. This could however tie in with what the new license report of Nuitka does, since it doesn't include DLL licenses coming from Python, or the license of libpython itself. Not sure where would be a good place to pick these up.

When I just checked my CPython3.10 installation on Windows, which I think has readline in its REPL, I failed to find it as a DLL, and I failed to find it as a mention in the license text of Python.

Ah, just found this: https://pypi.org/project/readline/

So, on Linux actually the problem exists. We do not include libc, and graphics drivers, since they are supposed to come from the OS. Added libreadline there would probably be good enough, however I fear that older systems and newer systems do not share the binary.

The Nuitka container build of Nuitka commercial is using a self-built Python without readline support, because that is bloat for most people anyway.

So I feel like, this is the responsibility of the one making the deployment. Since Linux deployment does not happen on new systems anyway, I guess making my Linux container build public, or instructions to that end, or to recommend Nuitka-Python on Linux, or something like that should be a solution. But honestly, I do not want to carry the burden of warning users about this. 

Loading a GPLed DLL that you do not use, in my mind doesn't trigger the GPL license, and honestly Python upstream should solve the issue by using another library, or even some pure Python.

**Comment by tropxy** [comment]
I am also not a lawyer, but I have dived quite deep in this subject, so take my comments and suggestions accordingly.

> Are you saying, that say using rar to pack/unpack a GPLed binary has any effect on the other binaries contained? The LGPL would indeed come into play if it was statically linked.

yes, statically or dynamically linking your code to  GPL library contaminates your License and you may forced to disclose your proprietary code that is using open source code.

check also these links:
* https://github.com/conda-forge/python-feedstock/issues/387
* https://velovix.github.io/post/lgpl-gpl-license-compliance-with-pyinstaller/

> The Nuitka container build of Nuitka commercial is using a self-built Python without readline support, because that is bloat for most people anyway.

if that is the case, then I think we are good here.

> So I feel like, this is the responsibility of the one making the deployment. Since Linux deployment does not happen on new systems anyway, I guess making my Linux container build public, or instructions to that end, or to recommend Nuitka-Python on Linux, or something like that should be a solution. But honestly, I do not want to carry the burden of warning users about this.

I would say that just a note of warning in the docs, like a FYI of how Nuitka addresses this issue would suffice.

> Loading a GPLed DLL that you do not use, in my mind doesn't trigger the GPL license, and honestly Python upstream should solve the issue by using another library, or even some pure Python.

I think it is a gray area and subject of a legal discussion. Python itself is GPL compatible, which means you can distribute a version of python with GPL libs and doesnt trigger the disclose source clause: 
https://docs.python.org/3/license.html

However, the GPL is clear and says if you include in your bundle a GPL lib, the clause is triggered. As the subject gets complicated, I assume that is why Python included options to remove readline or substitute it by editline which is under a LGPL license: 
https://docs.python.org/3/using/configure.html#cmdoption-with-readline

Thus, the safest option to remove any shred of doubt is really to avoid readline in packaging and distribution.

> So, on Linux actually the problem exists. We do not include libc, and graphics drivers, since they are supposed to come from the OS.

Btw, libc is under LGPL and as far as I remember, they added a special clause that allows binaries to dynamically or statically link to the lib without triggering the disclose source. So AFAIK, libc is never an issue.

**Comment by kayhayen** [comment]
What I meant with my libc comment, is that we do not ship with nuitka standalone these binaries, that are coming from the system, often in fact have to come from the system. If we could expect libreadline to be on target systems, we could not include it. However, Nuitka cannot change the upstream Python dependencies. That means, if `libreadline.so` is required by libpython, then there is nothing it can do. It surely is not a tool to educate users about the licencse of the distribution Python.

However, now I checked standalone mode on Ubuntu Python, and I couldn't find any libreadline dependency of libpython, which also doesn't make sense to me anyway. It's only a separate extension module called `readline.so`.

I don't appreciate your stance to imply you know more than others. I was e.g. a frequent reader of Groklaw back on the SCO vs Linux days. I believe I know this stuff. That doesn't make me an expert. 

But I somehow find it hard to bebelieve that the Ubuntu packag of Python is under GPL because it contains `readline.so`, and that this is the license I now have to apply when using it. In fact, readline is mentioned in Ubuntu as licensed under GPL with a pointer to the license. in e.g. `/usr/share/doc/python3.10/copyright`

Technically Nuitka onefile and Debian package bare no difference. Both have open code to pack and unpack the files on a target system.

Did Ubuntu/Debian forget to tell me that their Python package as a whole is now subject to the GPL terms? That is a tall ask to convince me of that. Whenever I learn something about licensing, it is ususally from there.


That said, let me state the facts that I have absorded so far.

1- Problem only for Linux Python. 
2. Upstreams may or may not be working to resolve it from their side. But since my users do a lot of old Python, with 2.6 still being supported, that's not really relevant.
3. Including readline is probably unnecessary for most software.
4. The readline.so is under GPL and recepients must be informed of that at least. 

Actions that might be taken as a result.

1. The ´anti-bloat` plugin might warn about including readline, and allow suppression of imports to it. Unless allowed, e.g. these kind of changes should be given the usual treatment.

```
/usr/lib/python3.10$ grep -re "import readline"
pdb.py:            import readline
asyncio/__main__.py:        import readline  # NoQA
code.py:            import readline
rlcompleter.py:    import readline
pstats.py:        import readline
site.py:            import readline
cmd.py:                import readline
cmd.py:                    import readline
cmd.py:            import readline
```

The anti-bloat warning page can point out for `readline` why including it might be particularily unwanted. But I am not going to create this myself. But since there is prior art, basically just somebody inclined needs to do it.

I think I likely excluded these parts of stdlib from being automatically included, although I am a bit unsure if `code.py`  can cover it. It might be the case, that dependencies are just not handled in stdlib that is automatically included, and that is why it's not necessary.

Since `readline` is not available on some platforms, it seems to me the code will tolerate replacements with `raise ImportError` pretty well usually Our goal is to eventually compile stdlib as well, and then it will make a difference, and in fact for some it may already be bad. My test case might have been too small.

2. The license report should not only include distributions, but also for parts of stdlib like `readline` the license information.

I welcome help on both. I do not see myself doing these actions any time soon. It will be enough if i am willing to carry the maintenance burden of the anti-bloat work for `readline`. However, it seems for Linux users, there might even be a benefit in this for reduced size and runtime performance.

**Comment by kayhayen** [comment]
I switched tags to express, that it's no longer something I didn't do. I did in fact now do enough research, but it doesn't mean I will attach a milestone, since it's not something I will implement myself, therefore help wanted. The problem is clearly that of the users, so I think it's their task or that of interested parties to contribute the 2 changes I am asking there.

**Comment by tropxy** [comment]
> What I meant with my libc comment, is that we do not ship with nuitka standalone these binaries, that are coming from the system, often in fact have to come from the system. If we could expect libreadline to be on target systems, we could not include it. However, Nuitka cannot change the upstream Python dependencies. That means, if `libreadline.so` is required by libpython, then there is nothing it can do. It surely is not a tool to educate users about the licencse of the distribution Python.
> 
> However, now I checked standalone mode on Ubuntu Python, and I couldn't find any libreadline dependency of libpython, which also doesn't make sense to me anyway. It's only a separate extension module called `readline.so`.
> 
> I don't appreciate your stance to imply you know more than others. I was e.g. a frequent reader of Groklaw back on the SCO vs Linux days. I believe I know this stuff. That doesn't make me an expert.
> 
> But I somehow find it hard to bebelieve that the Ubuntu packag of Python is under GPL because it contains `readline.so`, and that this is the license I now have to apply when using it. In fact, readline is mentioned in Ubuntu as licensed under GPL with a pointer to the license. in e.g. `/usr/share/doc/python3.10/copyright`
> 
> Technically Nuitka onefile and Debian package bare no difference. Both have open code to pack and unpack the files on a target system.
> 
> Did Ubuntu/Debian forget to tell me that their Python package as a whole is now subject to the GPL terms? That is a tall ask to convince me of that. Whenever I learn something about licensing, it is ususally from there.
> 
> That said, let me state the facts that I have absorded so far.
> 
> 1- Problem only for Linux Python. 2. Upstreams may or may not be working to resolve it from their side. But since my users do a lot of old Python, with 2.6 still being supported, that's not really relevant. 3. Including readline is probably unnecessary for most software. 4. The readline.so is under GPL and recepients must be informed of that at least.
> 
> Actions that might be taken as a result.
> 
> 1. The ´anti-bloat` plugin might warn about including readline, and allow suppression of imports to it. Unless allowed, e.g. these kind of changes should be given the usual treatment.
> 
> ```
> /usr/lib/python3.10$ grep -re "import readline"
> pdb.py:            import readline
> asyncio/__main__.py:        import readline  # NoQA
> code.py:            import readline
> rlcompleter.py:    import readline
> pstats.py:        import readline
> site.py:            import readline
> cmd.py:                import readline
> cmd.py:                    import readline
> cmd.py:            import readline
> ```
> 
> The anti-bloat warning page can point out for `readline` why including it might be particularily unwanted. But I am not going to create this myself. But since there is prior art, basically just somebody inclined needs to do it.
> 
> I think I likely excluded these parts of stdlib from being automatically included, although I am a bit unsure if `code.py` can cover it. It might be the case, that dependencies are just not handled in stdlib that is automatically included, and that is why it's not necessary.
> 
> Since `readline` is not available on some platforms, it seems to me the code will tolerate replacements with `raise ImportError` pretty well usually Our goal is to eventually compile stdlib as well, and then it will make a difference, and in fact for some it may already be bad. My test case might have been too small.
> 
> 2. The license report should not only include distributions, but also for parts of stdlib like `readline` the license information.
> 
> I welcome help on both. I do not see myself doing these actions any time soon. It will be enough if i am willing to carry the maintenance burden of the anti-bloat work for `readline`. However, it seems for Linux users, there might even be a benefit in this for reduced size and runtime performance.

> I don't appreciate your stance to imply you know more than others. I was e.g. a frequent reader of Groklaw back on the SCO vs Linux days. I believe I know this stuff. That doesn't make me an expert.

Hey hey...I didnt say that. What I meant was, I am not a lawyer, I did my homework, but given the fact I am not a lawyer, take my claims with that in mind and not as a golden bullet proof comment

**Comment by kayhayen** [comment]
Ok, I misread your statement. My bad, I should have auto-corrected to the obvious thing to say, even if it was not clear.

There is a problem if you mix *incompatible* licenses. Free Software licenses are by far not all compatible with each other, There is a list of GPL incompatible Free Software licenses for good reasons. the statement you made is misleading.

When the GPL comes into effect, the resulting package, say the python3.10 Debian package, would come under GPL. and Debian/Ubuntu would document that. But it is not. How come? I am 100% convinced, that Debian and Ubuntu do believe that the GPL from extension modules does not extend to Python from the fact, that the extension module is included. 

Mind you, if all your software is under MIT or public domain or whatever, and you use GPL code in there as well, forget about the license you put there, the resulting binaries are under GPL. You code also is licensed under GPL as well suddenly. That not happening to Python is a big clue. It being "Open Source" has nothing to do with it. That doesn't protect against the GPL in what they call virality.

So, how would a program become a dervied work of readline then. Including it and barely using it, should be extremely unlikely to cause it. A Qt program that includes `readline.so`  is no way under GPL. A program that makes heavy use of the module though, would definitely qualify.

So, personally i am not convinced this is a general problem the way you described it at all. I am also convinced that plenty of professional people will not even come close to this kind of "not a problem", and I respect them. These are the kind of people who care about the newly added license report and therefore it would be good for that to include it.

So, lets conclude it here. Without this coming from a real lawyer as a definitive negation, I am going to hold the following beliefs.

a) Including `readline.so`  can be done legally and without affecting the license of the remainder of the bundle.
b) It makes sense to avoid including `readline.so` without need or by accident.

Anything that is not related to indiviidual beliefs about a) and not to the steps I outlined for b) is only to increase the urgency of a) but I don't see any need. I outlined steps to take that fully solve the issue.

But I made it clear, that I will only accept others to work on this, but not do this myself unless of course somebody with money steps up to pay for it. But I guess, these people will just use a properly compiled Python without `readline` module. 

if nobody steps up to execute the anti-bloat work and the license report addition, I do not accept a further discussion as a replacement for that lack. From my point of view, this part is decided.
